
# Assignment 1: Predicting overall The Human Freedom Index

This notebook contains a set of exercises that will guide you through the different steps of this assignment. Solutions need to be code-based, _i.e._ hard-coded or manually computed results will not be accepted. Remember to write your solutions to each exercise in the dedicated cells and to not modify the test cells. When you are done completing all the exercises submit this same notebook back to moodle in **.ipynb** format.

<div class="alert alert-success">

The <a href="https://www.cato.org/human-freedom-index/2021 ">Human Freedom Index</a> measures economic freedoms such as the freedom to trade or to use sound money, and it captures the degree to which people are free to enjoy the major freedoms often referred to as civil liberties—freedom of speech, religion, association, and assembly— in the countries in the survey. In addition, it includes indicators on rule of law, crime and violence, freedom of movement, and legal discrimination against same-sex relationships. We also include nine variables pertaining to women-specific freedoms that are found in various categories of the index.

<u>Citation</u>

Ian Vásquez, Fred McMahon, Ryan Murphy, and Guillermina Sutter Schneider, The Human Freedom Index 2021: A Global Measurement of Personal, Civil, and Economic Freedom (Washington: Cato Institute and the Fraser Institute, 2021).
    
</div>

<div class="alert alert-danger"><b>Submission deadline:</b> Sunday, January 29th, 23:55</div>


In [1]:
import pandas as pd
import numpy as np

<div class="alert alert-info"><b>Exercise 1</b>

Load the Human Freedom Index data from the link: https://github.com/jnin/information-systems/raw/main/data/hfi_cc_2021.csv in a DataFrame called ```df```.

<br><i>[0.25 points]</i>
</div>
<div class="alert alert-warning">
Do not download the dataset. Instead, read the data directly from the provided link
</div>

In [2]:
# We read data directly from the github csv file, and call the head to have a first look.

df=pd.read_csv("https://github.com/jnin/information-systems/raw/main/data/hfi_cc_2021.csv")
df.head()

year  countries  ISO                         region  hf_score  hf_rank  \
0  2019    Albania  ALB                 Eastern Europe      8.14     43.0   
1  2019    Algeria  DZA     Middle East & North Africa      5.26    154.0   
2  2019     Angola  AGO             Sub-Saharan Africa      6.09    129.0   
3  2019  Argentina  ARG  Latin America & the Caribbean      7.38     74.0   
4  2019    Armenia  ARM        Caucasus & Central Asia      8.20     40.0   

   hf_quartile  pf_rol_procedural  pf_rol_civil  pf_rol_criminal  ...  \
0          2.0               5.97          4.76             4.26  ...   
1          4.0               5.21          5.64             4.35  ...   
2          4.0               2.72          4.43             3.60  ...   
3          2.0               6.83          5.94             4.35  ...   
4          1.0                NaN           NaN              NaN  ...   

   ef_regulation_business_adm  ef_regulation_business_bureaucracy  \
0                        5.65                                6.67   
1                        4.22                                2.22   
2                        2.94                                2.44   
3                        2.71                                5.78   
4                        5.17                                5.56   

   ef_regulation_business_start  ef_regulation_business_bribes  \
0                          9.74                           6.24   
1                          9.31                           2.58   
2                          8.73                           4.70   
3                          9.58                           6.53   
4                          9.86                           6.96   

   ef_regulation_business_licensing  ef_regulation_business_compliance  \
0                              5.62                               7.18   
1                              8.77                               7.03   
2                              7.92                               6.78   
3                              5.73                               6.51   
4                              9.30                               7.04   

   ef_regulation_business  ef_regulation  ef_score  ef_rank  
0                    6.85           7.70      7.81     31.0  
1                    5.69           5.84      4.90    162.0  
2                    5.59           5.97      5.50    153.0  
3                    6.14           5.99      5.50    153.0  
4                    7.32           7.82      8.03     15.0  

[5 rows x 125 columns]

In [3]:
#LEAVE BLANK


<div class="alert alert-info"><b>Exercise 2</b>

First write the code to drop all the columns from the DataFrame ```df``` except ```['hf_quartile', 'ef_regulation',  'pf_expression', 'region']```, then drop all the rows from ```df``` containing missing values present in the selected columns.

<br><i>[0.25 points]</i>
</div>

<div class="alert alert-warning">

Remember, Python is case-sensitive. The resulting DataFrame ```df``` should contain only four columns.

</div>


We can drop all missing values using pandas ```.dropna()``` function. This step should normally be done 
after train-test-split, but being explicitly asked now, we'll do it outright. 
We also display the before and after code by running the cell to appreciate the changes.


In [4]:
# We directly call the columns needed from our dataset.

df=df[['hf_quartile', 'ef_regulation',  'pf_expression', 'region']]

# We drop the missing values, and check the state of NaN before and after elimination.

print('Missing values before elimination!                                                                             ',
      df.isna().sum().sort_values())

df=df.dropna()

print("\n", 'Missing values after elimination!                                                                             ',
      df.isna().sum().sort_values())

Missing values before elimination!                                                                              pf_expression      0
region             0
ef_regulation    100
hf_quartile      113
dtype: int64

 Missing values after elimination!                                                                              hf_quartile      0
ef_regulation    0
pf_expression    0
region           0
dtype: int64


In [5]:
#LEAVE BLANK

<div class="alert alert-info"><b>Exercise 3</b> 
    
Write the code to create the feature matrix ```X``` (```ef_regulation```,  ```pf_expression```, and ```region```) and the target array ```y``` (```hf_quartile```), then split them into separate training and test sets with the relative size of 0.75 and 0.25. Store the training and tests feature matrix in variables called ```X_train``` and ```X_test```, and the training and test label arrays as ```y_train``` and ```y_test```.
    
<br><i>[1 point]</i>
</div>


**Checking the target variable type and distribution**

In order to inform our train_test_split approach, we perform a simple exploratory data analysis
for knowing target variable type (we'll see that being "quartile" a discrete float variable, we can
consider it a classification task). We'll also analyze how many different values we have in the target
feature ("hf_quartile") and its proportions, which seem to be balanced.


In [6]:
print("These are the types of our different features in our dataframe                                                     ",
      df.dtypes)

print('\n', 'These are the different values distribution in the target variable                                            ', 
      df["hf_quartile"].value_counts(normalize=True))

These are the types of our different features in our dataframe                                                      hf_quartile      float64
ef_regulation    float64
pf_expression    float64
region            object
dtype: object

 These are the different values distribution in the target variable                                             4.0    0.253883
3.0    0.250134
2.0    0.248527
1.0    0.247456
Name: hf_quartile, dtype: float64


**Creating our target feature Series and our features matrix**

We create target array and features matrix. We also perform the ```train_test_split```. We don't put a random_state parameter as we have not been instructed to do so, although it will be done in Exercise 7 to help make results repeatable downstream.

In [7]:
# We create target array and features marix.

X=df.drop("hf_quartile", axis=1)
y=df["hf_quartile"]

# We perform the train-test-split with the specified parameters.

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test=train_test_split(X, y,train_size=0.75, test_size=0.25)

In [8]:
# LEAVE BLANK

In [9]:
# LEAVE BLANK

<div class="alert alert-info"><b>Exercise 4 </b> 

    
The resulting feature matrix contains a categorical variable. Write the code to create a ```ColumnTransformer``` to encode it using the one-hot encoding method. Store the transformer in a variable called ```transformer```. At this stage, you do not need to run it.

<br><i>[1 points]</i>
</div>

<div class='alert alert-warning'>

Not all the attributes are categorical. Ensure that all non-categorical attributes remain intact.
</div>

**ColumnTransformer: OneHotEncoding for "region" variable**

As we saw in the previous df.dtypes exploratory analysis, "region" feature is the categorical feature.
We will transform it with a ```ColumnTransformer```, putting ```OneHotEncoder``` class as the transformation. We then use the parameter ```remainder="passthrough``` to leave the remaining X_train variables untouched. In order to call the index position of "region" inside ```X_train```, we used ```X_train.columns.get_loc("region")```.

In [10]:
# We import the necessary scikit-learn classes that we will use in the ColumnTransformer and the Pipeline.
# These classes will also be used in Exercise 7.

from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline

In [11]:
#We create the transformer with the ColumnTransformer, complying with the guidelines.

transformer=ColumnTransformer([("ohe", OneHotEncoder(sparse=False), [X_train.columns.get_loc("region")])],
                              remainder="passthrough")

In [12]:
# LEAVE BLANK

In [13]:
# LEAVE BLANK

In [14]:
# LEAVE BLANK

<div class="alert alert-info"><b>Exercise 5 </b> 

Write the code to create a ```Pipeline``` consisting of a ```SimpleImputer``` with the most frequent strategy, the previous transformer, a standard scaler, and a logistic regression model. Store the resulting pipeline in a variable called ```pipe```.
    
<br><i>[1.5 points]</i>
</div>

<div class='alert alert-warning'>

Be sure you apply the data transformations in the correct order.
</div>

We instantiate the class ```Pipeline``` by means of the creation of a list of steps that we will use as parameters. Then, automation with transformation in it is feasible. We understand that having dropped all NaN's, the SimpleImputer could be skipped.

In [15]:
# We create a list with the subsequent instantiation of the different classes needed in the pipeline.

steps=[("Imputation", SimpleImputer(strategy="most_frequent")), 
       ("Transformer", transformer), 
       ("Scaler", StandardScaler()),
       ("Logistic Regression", LogisticRegression())]

# Lastly, we instantiate the pipeline with the steps.

pipe=Pipeline(steps)

In [16]:
# LEAVE BLANK

In [17]:
# LEAVE BLANK

In [18]:
# LEAVE BLANK

In [19]:
# LEAVE BLANK

In [20]:
# LEAVE BLANK

<div class="alert alert-info"><b>Exercise 6 </b> 
    
Write the code to store the achieved ```score``` (accuracy) in a variable called ```score```. 
    
<br><i>[1 point]</i>
</div>

<div class='alert alert-warning'>

Use train and test datasets correctly.
</div>

We train/fit the pipeline as if it was a normal ML model with our training data, and then we proceed to calculate the scores with our test data (scores are built-in inside the pipeline). 

In [21]:
# We fit the pipeline as if it was any other model, with the train data. We predict with our test data.

pipe.fit(X_train, y_train)
pipe.predict(X_test)

# We output accuracy with test data.

score= pipe.score(X_test, y_test)
print(score)

0.7644539614561028


In [22]:
# LEAVE BLANK

<div class="alert alert-info"><b>Exercise 7 </b> 
    
The previous exercises were simple because they included only three columns. Now repeat the same process but using the complete dataset. This exercise is open. You can use any scaler, imputer, transformer, or encoder. The only requirement is to train a logistic regression. If you decide to drop a column, justify the reason. 
    
<br><i>[5 points]</i>
</div>

<div class='alert alert-warning'>
    
The following columns are redundant and should be dropped:
* ```year```
* ```ISO```
* ```countries```
* All columns containing the word ```rank``` 
* All columns containing the word ```score```
    
</div>


## Outline of the Exercise 7

We have structured this exercise in seven tasks:

1. Elimination of unique identifiers and redundant variables
2. Multicollinearity: removal of features with Pearson > 90%
3. Missing values in target feature: removal
4. Missing values > 5%: feature removal
5. Target and features data + train-test-split
6. ColumnTransformer and Pipeline
7. Amplify evaluation metrics: Confusion matrix + Precision and Recall

All necessary classes already imported in previous exercises will not be re-stated if not new.

In [23]:
# We call again our whole dataframe to perform needed operations from scratch.

df=pd.read_csv("https://github.com/jnin/information-systems/raw/main/data/hfi_cc_2021.csv")
df.head()

year  countries  ISO                         region  hf_score  hf_rank  \
0  2019    Albania  ALB                 Eastern Europe      8.14     43.0   
1  2019    Algeria  DZA     Middle East & North Africa      5.26    154.0   
2  2019     Angola  AGO             Sub-Saharan Africa      6.09    129.0   
3  2019  Argentina  ARG  Latin America & the Caribbean      7.38     74.0   
4  2019    Armenia  ARM        Caucasus & Central Asia      8.20     40.0   

   hf_quartile  pf_rol_procedural  pf_rol_civil  pf_rol_criminal  ...  \
0          2.0               5.97          4.76             4.26  ...   
1          4.0               5.21          5.64             4.35  ...   
2          4.0               2.72          4.43             3.60  ...   
3          2.0               6.83          5.94             4.35  ...   
4          1.0                NaN           NaN              NaN  ...   

   ef_regulation_business_adm  ef_regulation_business_bureaucracy  \
0                        5.65                                6.67   
1                        4.22                                2.22   
2                        2.94                                2.44   
3                        2.71                                5.78   
4                        5.17                                5.56   

   ef_regulation_business_start  ef_regulation_business_bribes  \
0                          9.74                           6.24   
1                          9.31                           2.58   
2                          8.73                           4.70   
3                          9.58                           6.53   
4                          9.86                           6.96   

   ef_regulation_business_licensing  ef_regulation_business_compliance  \
0                              5.62                               7.18   
1                              8.77                               7.03   
2                              7.92                               6.78   
3                              5.73                               6.51   
4                              9.30                               7.04   

   ef_regulation_business  ef_regulation  ef_score  ef_rank  
0                    6.85           7.70      7.81     31.0  
1                    5.69           5.84      4.90    162.0  
2                    5.59           5.97      5.50    153.0  
3                    6.14           5.99      5.50    153.0  
4                    7.32           7.82      8.03     15.0  

[5 rows x 125 columns]

### 1. Elimination of unique identifiers and redundant variables ###

We eliminate the aforementioned variables from the original dataframe, as they are either unique identifiers without predictive power or redundant in terms of information. By such operations, we eliminated 11 features and are left with 114.

In [24]:
df.drop(["year", "ISO", "countries"], axis=1, inplace=True)
df.drop(list(df.filter(regex = 'rank')), axis = 1, inplace = True)
df.drop(list(df.filter(regex = 'score')), axis = 1, inplace = True)

print(f'We are left with {df.shape[1]} features from the original 125.')

We are left with 114 features from the original 125.


### 2. Multicollinearity: removal of features with Pearson > 90%

Secondly, we will check for multicollinearity. In order to simplify our analysis, we'll temporarily drop 
"region", the only non-numerical variable. Then, given the difficulty of plotting a Seaborn heatmap 
with 114 columns, we will proceed to use 0.9 as a correlation threshold to drop features already 
explained by others through detailed code. We use 0.9 because we want to follow a conservative approach and not drop features that may have good predictive power. We end up dropping 43 variables, and are left with 71.

In [25]:
# We drop first temporarily "region" to simplify our correlation analysis for multicollinearity spotting.

df1=df.drop("region", axis=1)

In [26]:
#We create the correlation matrix numerically without displaying it.

corr_matrix=df1.corr().round(2)

correlated_pairs = corr_matrix.abs().unstack().sort_values(ascending=False)

# get the pairs with correlation greater than a threshold (here, 0.9).
threshold = 0.9
correlated_pairs = correlated_pairs[correlated_pairs > threshold]

# drop the diagonal values, with perfect correlation with the own variable.
correlated_pairs = correlated_pairs[correlated_pairs != 1]

# display the correlated pairs
print(correlated_pairs)

pf_womens                          pf_identity                          0.98
pf_identity                        pf_womens                            0.98
pf_identity_inheritance            pf_identity_inheritance_daughters    0.98
                                   pf_identity_inheritance_widows       0.98
pf_identity_inheritance_daughters  pf_identity_inheritance              0.98
                                                                        ... 
pf_rol                             ef_legal_courts                      0.91
pf_expression                      pf_expression_cultural               0.91
                                   pf_assembly                          0.91
                                   pf_expression_media                  0.91
pf_expression_cultural             pf_expression                        0.91
Length: 70, dtype: float64


In [27]:
# Then, we treat this Series Index as a list, join them together and take out only those features that are
# unique, which can be carried out with the set() method.

correlated_variables=list(correlated_pairs.index)
correlated_variables

variables_repeated=[]
for pair in correlated_variables:
    variables_repeated.append(pair[0])
    variables_repeated.append(pair[1])

multicollinearity=list(set(variables_repeated))

print('These are correlated variables with Pearson >90%, having {} variables:'.format(len(multicollinearity)),
      "\n",multicollinearity)

These are correlated variables with Pearson >90%, having 43 variables: 
 ['ef_trade_regulatory', 'pf_movement', 'pf_assembly_parties_bans', 'pf_expression_cultural', 'pf_assembly_parties_barriers', 'pf_identity_same', 'pf_assembly_freedom_cld', 'pf_assembly', 'ef_trade_regulatory_compliance', 'pf_expression_freedom_bti', 'pf_expression_media', 'pf_expression', 'pf_identity_inheritance_daughters', 'pf_rol_civil', 'pf_identity', 'pf_assembly_freedom', 'pf_rol_procedural', 'pf_rol_criminal', 'pf_identity_inheritance', 'pf_assembly_freedom_house', 'ef_government_tax_payroll', 'pf_movement_cld', 'ef_government_tax', 'pf_expression_freedom', 'pf_identity_same_m', 'pf_assembly_freedom_bti', 'pf_religion_freedom_vdem', 'pf_religion_suppression', 'pf_religion_freedom', 'pf_assembly_civil', 'pf_assembly_parties', 'pf_expression_freedom_cld', 'pf_religion', 'pf_identity_inheritance_widows', 'pf_movement_vdem', 'pf_identity_same_f', 'pf_movement_vdem_foreign', 'pf_assembly_parties_auton', 'pf_reli

In [28]:
#We drop those 43 features that affect multicollinearity in the main DataFrame.

df.drop(multicollinearity, axis=1, inplace=True)

print(f'We drop {len(multicollinearity)} variables, ending up with {df.shape[1]}.')

We drop 43 variables, ending up with 71.


### 3. Missing Values in target Feature: removal

Logistic Regression does not accept missing values, and this holds true for the target feature. We could run a NaN analysis after ```train-test-split``` and either eliminate them or impute them inside a ```Pipeline```, but the target feature NaN's would trigger an error (as we don't touch them).

Therefore, as imputation in the target feature would not make any predictive sense (as we would alter the estimation of reality), we will first remove them before the split. We'll do it specifying the "subset" parameter, dropping 113 instances and finishing up with 1867.

In [29]:
df.dropna(subset=["hf_quartile"], inplace=True)

print(f'In total, we have dropped {1980-df.shape[0]} instances by removing target feature NaNs.')

In total, we have dropped 113 instances by removing target feature NaNs.


### 4. Missing Values > 5%: feature removal

We will now proceed to impute different missing values. We follow two approaches:

1. If missing data in a feature is >5% of all instances, we remove the feature. We do this to prevent additional noise.
2. If missing data in a feature is <5%, we replace with the median on numerical features, and with mode on categorical features. This step will be carried out after train_test_split.

We explore first the degree of missing values by feature, eliminate features with >5% of NaN, and finally create the imputers. At he end, we drop 5 columns and we are left with 65 features.

In [30]:
# We first create a DataFrame displaying the number of missing values and proportions per column.

import numpy as np

columns= list(df.columns)
missing=list(df.isna().sum())
proportionmiss=list((np.array(df.isna().sum())/1867).round(2))

missing_df=pd.DataFrame({"columns": columns, "missing": missing, "proportionnmiss": proportionmiss})
missing_df.sort_values("proportionnmiss", ascending=False)

columns  missing  proportionnmiss
30               ef_legal_military      175             0.09
63      ef_regulation_business_adm      135             0.07
57      ef_regulation_labor_firing      137             0.07
5   pf_ss_disappearances_organized      134             0.07
58     ef_regulation_labor_bargain      136             0.07
..                             ...      ...              ...
37                     ef_money_sd        0             0.00
38              ef_money_inflation        0             0.00
39               ef_money_currency        0             0.00
40                        ef_money        0             0.00
70                   ef_regulation        0             0.00

[71 rows x 3 columns]

In [31]:
# We have 6 variables that surpass the >5% of NaN threshold. We remove them from the DataFrame.

print(missing_df[missing_df["proportionnmiss"]>0.05])

missing_5=list(missing_df[missing_df["proportionnmiss"]>0.05]["columns"])
df.drop(missing_5, axis=1, inplace=True)

print("\n", f'In total, we have dropped {len(missing_5)} columns, ending up with {df.shape[1]} features.')

                           columns  missing  proportionnmiss
5   pf_ss_disappearances_organized      134             0.07
30               ef_legal_military      175             0.09
52  ef_regulation_credit_ownership      105             0.06
57      ef_regulation_labor_firing      137             0.07
58     ef_regulation_labor_bargain      136             0.07
63      ef_regulation_business_adm      135             0.07

 In total, we have dropped 6 columns, ending up with 65 features.


###  5. Target and features data + train_test_split

Up to here, we have the final data to input to our model. 

**Columns**: All highly correlated variables from a 0.9 of Pearson score are removed, and columns with NaN> 5% are also out, being left with 65 features.

**Instances**: First removal of target feature NaN's, as the final pipeline would output an error in the score.

We will now create our target array and features matrix, performing the ```train_test_split``` with a ```random_state``` to make our results repeatable downstream.

In [32]:
# We create our target array (y) and our features matrix (X) with a random_state.

X=df.drop("hf_quartile", axis=1)
y=df["hf_quartile"]

# We then proceed to split our data.

X_train, X_test, y_train, y_test=train_test_split(X,y, test_size=0.25, random_state=42)

### 6. ColumnTransformer and Pipeline ###

We will now set a column transformation strategy for each different category. The strategy for each type is:

1. Categorical features ("region"): We OneHotEncode the variable, no imputation needed as no missing values are present.
2. Numerical features: we replace missing values by median, with less sensitivity to extreme values tthan mean.

We will create for each typology the ColumnTransformer, and impute it inside the final pipeline that will add a scaler (as logistic regression is sensible in terms of distance) and the ```LogisticRegression``` model itself.

In [33]:
# First, we ensure that "region" is the only categorical variable in our train DataFrame. It is.

X_train.dtypes.value_counts()

float64    63
object      1
dtype: int64

In [34]:
# We create he ColumnTransformer with the aforementioned strategies for categorical and numerical values.
# We call by index positions the different columns, as strings are not supported for NumPy arrays.

imputer=SimpleImputer(strategy="median")
onehot=OneHotEncoder(sparse=False)

transformer=ColumnTransformer([("encode", onehot, [X_train.columns.get_loc("region")]), 
                              ("medianimpute", imputer, list(range(1,64)))]
                              ,remainder="passthrough")

In [35]:
# We now create the pipeline.

scaler=StandardScaler()
logreg=LogisticRegression()

steps=[("transformer", transformer),
       ("scaler", scaler),
       ("logreg", logreg)]

pipe=Pipeline(steps)

In [36]:
# We fit the model to our training data and calculate the accuracy score.

pipe.fit(X_train, y_train)
pipe.score(X_test, y_test)

/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


0.9100642398286938

### 7. Amplifying evaluation metrics: Confusion matrix + Precision and Recall ###

**The accuracy score of 91% shows an increase regarding the previous dataset (approximately, of +15%). Those are good news, but we can actually further breakdown prediction metrics by each of the four quartiles precision and recall.**

We will see that quartile 2 shows the worst general metrics, although overall, quality of predictions (precision) and sensitivity of predictions (recall) seem good, being always superior to 85%.

***Given such metrics, we assume the exercise can be concluded, at the expense of waiting for the business objective for fully determining if metrics are good enough or not.***

In [37]:
# We import the necessary modules for further evaluation.

from sklearn.metrics import classification_report, confusion_matrix

In [38]:
# We predict the instances in order to make a comparison with already established test features.

y_pred=pipe.predict(X_test)

# Evaluation metrics broken down by different categories, displaying also the confusion matrix.

print(confusion_matrix(y_test, y_pred))
print("\n", classification_report(y_test, y_pred))

[[110  10   0   0]
 [ 11 100   3   0]
 [  0   8 104   6]
 [  0   0   4 111]]

               precision    recall  f1-score   support

         1.0       0.91      0.92      0.91       120
         2.0       0.85      0.88      0.86       114
         3.0       0.94      0.88      0.91       118
         4.0       0.95      0.97      0.96       115

    accuracy                           0.91       467
   macro avg       0.91      0.91      0.91       467
weighted avg       0.91      0.91      0.91       467

